### PR TITLE
Add `racecar` integration

### PR DESCRIPTION
**Pull request by p-lambert**
This PR provides instrumentation support for [racecar](https://github.com/zendesk/racecar).

#### Usage

```rb
Datadog.configure do |c|
  c.use :racecar
end
```

### CODE REVIEW

File: lib/ddtrace/contrib/racecar/patcher.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+        module_function
+
+        def patch
+          return patched? if patched? || !defined?(::Racecar)
+
+          require_relative 'tracer'
+
+          ::Racecar.singleton_class.class_eval do
+            alias_method :__instrumenter, :instrumenter
```
Comment (dasch) [inline lib/ddtrace/contrib/racecar/patcher.rb]: Would it be possible to make this work now that the `start_` notifications are emitted?

File: lib/ddtrace/contrib/racecar/patcher.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        option :service_name, default: 'racecar'
+
+        class << self
+          def patch
+            return patched? if patched? || !compatible?
+
+            ::ActiveSupport::Notifications.subscribe('start_process_batch.racecar', &method(:start))
+            ::ActiveSupport::Notifications.subscribe('start_process_message.racecar', &method(:start))
+            ::ActiveSupport::Notifications.subscribe('process_batch.racecar', &method(:finish))
+            ::ActiveSupport::Notifications.subscribe('process_message.racecar', &method(:finish))
```
Comment (dasch) [inline lib/ddtrace/contrib/racecar/patcher.rb]: I think you need to separate message processing from batch processing – at least give the spans different names.

File: lib/ddtrace/contrib/racecar/patcher.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            ensure_clean_context!
+
+            span = configuration[:tracer].trace(NAME)
+            span.service = configuration[:service_name]
+            span.resource = payload[:consumer_class]
+            span.set_tag('kafka.topic', payload[:topic])
+            span.set_tag('kafka.consumer', payload[:consumer_class])
+            span.set_tag('kafka.partition', payload[:partition])
+            span.set_tag('kafka.offset', payload[:offset]) if payload.key?(:offset)
+            span.set_tag('kafka.first_offset', payload[:first_offset]) if payload.key?(:first_offset)
```
Comment (dasch) [inline lib/ddtrace/contrib/racecar/patcher.rb]: This mixes up batch processing and individual message processing. For the former, you'd be interested in

1. The start and end offsets
2. The batch size

For the latter you're just interested in the offset of that specific message.

File: lib/ddtrace/contrib/racecar/patcher.rb (hunk 2 above)
Comment (p-lambert) [reply lib/ddtrace/contrib/racecar/patcher.rb]: Hey @dasch! Apologies for the late reply. I'm not sure if I fully understand the problem here. As far as I understand each worker either implements a `process` or `process_batch` method, correct? So each span will belong to a resource that univocally represents either batch or a message worker.

File: lib/ddtrace/contrib/racecar/patcher.rb (hunk 3 above)
Comment (p-lambert) [reply lib/ddtrace/contrib/racecar/patcher.rb]: I'm conditionally setting the fields `kafka.offset` and `kafka.first_offset`.
* For the `process_message.racecar` event you provide a payload with `consumer_class`, `topic`,  `partition` and `offset` so these are set in the span accordingly.
* For the `process_batch.racecar`, event you provide a payload with `consumer_class`, `topic`,  `partition` and `first_offset` which are the fields set in batch consumer spans.
Let me know if that makes sense!

File: lib/ddtrace/contrib/racecar/patcher.rb (hunk 2 above)
Comment (dasch) [reply lib/ddtrace/contrib/racecar/patcher.rb]: That's true, but a service can change from one to the other, and often will – batching is a performance optimization. Naming the top-level span either `process_message` or `process_batch` would make it clear what the trace represents.

File: lib/ddtrace/contrib/racecar/patcher.rb (hunk 3 above)
Comment (dasch) [reply lib/ddtrace/contrib/racecar/patcher.rb]: Sure, I was just confused by both message and batch processing was mixed together. For batch processing, the number of messages is obviously important.

File: lib/ddtrace/contrib/racecar/patcher.rb (hunk 2 above)
Comment (p-lambert) [reply lib/ddtrace/contrib/racecar/patcher.rb]: I see! I'll make change that change then!

File: lib/ddtrace/contrib/racecar/patcher.rb (hunk 3 above)
Comment (p-lambert) [reply lib/ddtrace/contrib/racecar/patcher.rb]: So do you plan to update the instrumentation payload with that field?

Comment (dasch) [reply lib/ddtrace/contrib/racecar/patcher.rb]: Ah, thought I already had that! I'll add it now.

Comment (dasch) [reply lib/ddtrace/contrib/racecar/patcher.rb]: https://github.com/zendesk/racecar/pull/47

You can add the field conditionally, I'll merge this ASAP.

Comment (p-lambert) [reply lib/ddtrace/contrib/racecar/patcher.rb]: Cool, thank you!

**Pull request merged**